compiler/rustc_hir_typeck/src/pat.rs

```diff
@@ -2125,45 +2204,67 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     // the bad interactions of the given hack detailed in (note_1).
                     debug!("check_pat_ref: expected={:?}", expected);
                     match *expected.kind() {
-                        ty::Ref(_, r_ty, r_mutbl) if r_mutbl == mutbl => (expected, r_ty),
-                        _ => {
-                            if consumed_inherited_ref && self.tcx.features().ref_pat_everywhere {
-                                // We already matched against a match-ergonmics inserted reference,
-                                // so we don't need to match against a reference from the original type.
-                                // Save this infor for use in lowering later
-                                self.typeck_results
-                                    .borrow_mut()
-                                    .skipped_ref_pats_mut()
-                                    .insert(pat.hir_id);
-                                (expected, expected)
+                        ty::Ref(_, r_ty, r_mutbl) if r_mutbl == mutbl => {
+                            let pat_info = if r_mutbl == Mutability::Not
+                                && ((pat.span.at_least_rust_2024()
+                                    && self.tcx.features().ref_pat_eat_one_layer_2024)
+                                    || self.tcx.features().ref_pat_everywhere)
+                            {
+                                PatInfo { max_ref_mutbl: MutblCap::Not, ..pat_info }
```

Comment:
Could this be done in `calc_default_binding_mode`? I think you have all the info required there

Reply:
Hmm… `check_pat_ref` uses `shallow_resolve` to get the reference type, but other parts of the file use `try_structurally_resolve_type` instead. Do you know what the difference is? Does it matter which one I use?

(Also, I'm not sure moving to `calc_default_binding_mode` would be cleaner. Right now, `calc_default_binding_mode` manages inherited references and `check_pat_ref` manages non-inherited references, which is a reasonably clean separation. Maybe we could do a wider refactor where `check_pat_ref` assumes responsibility for all `&`/`&mut` pattern logic, but that is best left to a future PR I think.)

Reply:
I don't know the difference but I don't think it matters here. The reason I think it's better over on `calc_default_binding_mode` is that it's `max_ref_mutbl` capping happens there already so it would be easier to track in one place. In fact I think this one (line 2213) partially overrides the one on line 359

Reply:
>  a wider refactor where `check_pat_ref` assumes responsibility for all `&`/`&mut` pattern logic

I did this.

Reply:
That's great, ty